### PR TITLE
Feature/platfowner/cross shard

### PR DESCRIPTION
**Pull request by platfowner**
Summary:
- Add StateManager with skeleton code.
- Integrate backupDb, node db, and consensus logic with the state manager.
- Add tests for the state manager.
- Apply some renaming.
- Add enableStateVersionOpt feature flag and TODOs for the next step.

### CODE REVIEW

File: db/state-manager.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+} = require('./state-util');
+const {
+  FeatureFlags,
+  StateVersions,
+} = require('../constants');
+
+class StateManager {
+  constructor() {
+    this.rootMap = new Map();
+    this.setRoot(StateVersions.FINAL, new StateNode());
```
Comment (liayoo) [inline db/state-manager.js:15]: Seems like this.finalizedVersion is initialized to StateVersions.FINAL and its state is always an empty node? Could we rename this?

File: db/index.js [hunk 2]
```diff
@@ -26,27 +28,26 @@ const {
...
 const RuleUtil = require('./rule-util');
 
 class DB {
-  constructor(bc, tp, isFinalizedState, blockNumberSnapshot) {
+  constructor(stateRoot, stateVersion, bc, tp, isFinalizedState, blockNumberSnapshot) {
     this.shardingPath = null;
-    this.isRoot = null;
-    this.stateTree = new StateNode();
-    this.initDbData();
+    this.isRoot = null;  // Is this the database of the root blockchain?
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/index.js]: Done. Thanks.

File: consensus/index.js [hunk 3]
```diff
@@ -560,15 +581,17 @@ class Consensus {
       // FIXME: ask for the block from peers
       return false;
     }
-    const tempState = this.getStateSnapshot(block);
-    if (!tempState) {
+    const snapVersion = `${StateVersions.CONSENSUS}:snapshot:${Date.now()}`;
```
[the remark this answers is not in the document]
Comment (platfowner) [reply consensus/index.js]: Moved it to the inside of getSnapDb(). Thanks.

File: db/state-manager.js (hunk 1 above)
Comment (platfowner) [reply db/state-manager.js]: Renamed it StateVersions.EMPTY, and clarified finalzeVersion(). Thanks.

**Comment by platfowner** [comment]
All comments addressed. PTAL @liayoo

File: server/index.js [hunk 4]
```diff
@@ -33,6 +33,7 @@ const {
   FunctionProperties,
   FunctionTypes,
   NativeFunctionIds,
+  StateVersions,
```
Comment (liayoo) [inline server/index.js:36]: unused?

File: node/index.js [hunk 5]
```diff
@@ -61,18 +67,54 @@ class BlockchainNode {
...
     this.initialized = true;
     return lastBlockWithoutProposal;
   }
 
+  createTempDb(baseVersion, newVersion, blockNumberSnapshot) {
+    return this.createDb(baseVersion, newVersion, null, null, false, blockNumberSnapshot);
+  }
+
+  createDb(baseVersion, newVersion, bc, tp, isFinalizedState, blockNumberSnapshot) {
+    const newRoot = this.stateManager.cloneVersion(baseVersion, newVersion);
```
Comment (liayoo) [inline node/index.js:90]: do we need a null check?

File: node/index.js [hunk 6]
```diff
@@ -61,18 +67,54 @@ class BlockchainNode {
...
+  }
+
+  syncDb(newVersion) {
+    const LOG_HEADER = 'syncDb';
+    const oldVersion = this.db.stateVersion;
+    if (newVersion === oldVersion) {
+      logger.info(`[${LOG_HEADER}] Already sync'ed.`);
+      return false;
+    }
+    const clonedRoot = this.stateManager.cloneFinalizedVersion(newVersion);
```
Comment (liayoo) [inline node/index.js:108]: ditto

**Review by liayoo — APPROVED** [review]
LGTM with some comments. Thanks for the groundwork! 👍

File: server/index.js (hunk 4 above)
Comment (platfowner) [reply server/index.js]: Removed. Thanks.

File: node/index.js (hunk 5 above)
Comment (platfowner) [reply node/index.js:91]: Done. Thanks.

File: node/index.js (hunk 6 above)
Comment (platfowner) [reply node/index.js:113]: ditto

**Comment by platfowner** [comment]
All comments address! Now merging..

**Pull request merged**